lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
+/// each task. The typical task is a Swift frontend job.
+/// Maintain invarients connecting tasks to execute, tasks currently executing,
+/// and fds being polled. Handle output from and death of tasks.
+class TaskShepherd {
+
+  /// The set of tasks needed to be executed.
+  std::queue<std::unique_ptr<Task>> &TasksToExecute;
+
+public:
+  /// Unix system calls deal in process IDs (Pids), so a means is needed
+  /// to map a Pid back to a Task structure in memory.
   using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
 
-  // Stores the current executing Tasks, organized by pid.
-  PidToTaskMap ExecutingTasks;
+private:
+  TaskMap ExecutingTasks;
 
   // Maintains the current fds we're checking with poll.
-  std::vector<struct pollfd> PollFds;
+  std::vector<struct pollfd> FdsToPoll;
 
-  bool SubtaskFailed = false;
+  /// Limits the number of tasks to run in parallel
+  const unsigned MaxNumberOfParallelTasks;
 
-  unsigned MaxNumberOfParallelTasks = getNumberOfParallelTasks();
+  /// Optional functions to call to mark significant events in a Task's life.
+  const TaskQueue::TaskBeganCallback BeganCallBack;
+  const TaskQueue::TaskFinishedCallback FinishedCallBack;
+  const TaskQueue::TaskSignalledCallback SignalledCallBack;
 
-  if (MaxNumberOfParallelTasks == 0)
-    MaxNumberOfParallelTasks = 1;
+  /// Optional place to count I/O and subprocess events.
+  UnifiedStatsReporter *Stats;
 
-  while ((!QueuedTasks.empty() && !SubtaskFailed) ||
-         !ExecutingTasks.empty()) {
-    // Enqueue additional tasks, if we have additional tasks, we aren't
-    // already at the parallel limit, and no earlier subtasks have failed.
-    while (!SubtaskFailed && !QueuedTasks.empty() &&
-           ExecutingTasks.size() < MaxNumberOfParallelTasks) {
-      std::unique_ptr<Task> T(QueuedTasks.front().release());
-      QueuedTasks.pop();
-      if (T->execute())
-        return true;
+public:
+  TaskShepherd(std::queue<std::unique_ptr<Task>> &TasksToExecute,
+               const unsigned NumberOfParallelTasks,
+               const TaskQueue::TaskBeganCallback BeganCallBack,
+               const TaskQueue::TaskFinishedCallback FinishedCallBack,
+               const TaskQueue::TaskSignalledCallback SignalledCallBack,
+               UnifiedStatsReporter *Stats)
+      : TasksToExecute(TasksToExecute),
+        MaxNumberOfParallelTasks(
+            NumberOfParallelTasks == 0 ? 1 : NumberOfParallelTasks),
+        BeganCallBack(BeganCallBack), FinishedCallBack(FinishedCallBack),
+        SignalledCallBack(SignalledCallBack), Stats(Stats) {}
+
+  /// Run the tasks in the queue, \return true on error.
+  bool executeTasks();
+
+private:
+  bool isFinishedExecutingTasks() const {
+    return ExecutingTasks.empty() && TasksToExecute.empty();
+  }
 
-      pid_t Pid = T->getPid();
+  /// Start up tasks if we aren't
+  /// already at the parallel limit, and no earlier subtasks have failed.
+  /// \return true for error;
```

Comment:
Was there more to come after this semicolon?

Reply:
Turns out you reviewed an early version of this PR (sorry)--I have already cleaned this up as part of more reorganization. Will let you know when it is ready for another pass.